libutils/glob_lib.c

```diff
@@ -0,0 +1,588 @@
+#include <platform.h>
+#include <glob_lib.h>
+#include <alloc.h>
+#include <file_lib.h>
+#include <string_lib.h>
+#include <logging.h>
+#include <sequence.h>
+#include <path.h>
+
+/**
+ * If not compiled with PCRE2, we fallback to glob(3) implementation. Why not
+ * use glob(3) and fallback on our implementation? Because commonly used
+ * features like brace expansion is not defined by POSIX, and would require
+ * custom implementation anyways. We don't want subtle differences in features
+ * for different platforms.
+ */
+#ifdef WITH_PCRE2
+#include <regex.h>
+#else // WITH_PCRE2
+#include <glob.h>
+#endif // WITH_PCRE2
+
+/**
+ * On Windows we accept both forward- & backslash, otherwise only forward slash
+ * is accepted.
+ */
+#ifdef _WIN32
+#define PATH_DELIMITERS "\\/"
+#else // _WIN32
+#define PATH_DELIMITERS "/"
+#endif // _WIN32
+
+/**
+ * Returns a set of brace-expanded patterns.
+ */
+static void ExpandBraces(const char *pattern, StringSet *expanded)
+    FUNC_UNUSED; // Unused if compiled without PCRE2, but tested in unit tests
+
+static void ExpandBraces(const char *const pattern, StringSet *const expanded)
+{
+    assert(pattern != NULL);
+    assert(expanded != NULL);
+
+    char *const start = SafeStringDuplicate(pattern);
+
+    /* First we will split the pattern into three parts based on the curly
+     * brace delimiters. We look for the shortest possible match. E.g., the
+     * string "foo{{bar,baz}qux,}" becomes "foo{", "bar,baz", "qux,}".
+     */
+    char *left = NULL, *right = NULL;
+    for (char *ch = start; *ch != '\0'; ch++)
+    {
+        if (*ch == '{')
+        {
+            left = ch;
+        }
+        else if (left != NULL && *ch == '}')
+        {
+            right = ch;
+            break;
+        }
+    }
+
+    /* Next we check if base case is reached. I.e., if there is no curly brace
+     * pair to expand, then add the pattern itself to the result.
+     */
+    if (right == NULL)
+    {
+        StringSetAdd(expanded, start);
+        return;
+    }
+    *left = '\0';
+    *right = '\0';
+
+    /* Next we split the middle part (between the braces) on the comma
+     * delimiter. E.g., the string "bar,baz" becomes "bar", "baz".
+     */
+    char *middle = left + 1, *end = right + 1;
+    Seq *const split = StringSplit(middle, ",");
+
```

Comment:
What happens here if there is no comma? Worth an assert / early return / comment.